# Taller Pandas 


Para el siguiente taller se trabajaran con los archivos mencionados en cada ejercicio. Asegurarse de tenerlos en la misma carpeta que este documento.

Taller basado en el curso __Introduction to Data Science in Python__ de la universidad de Michigan.

## Parte 1

Importe el archivo olympics.csv. Asegurese que quede en el formato adecuado.

In [3]:
import pandas as pd

In [232]:
df = pd.read_csv('olympics.csv',header=1, index_col=0)
df.head()

№ Summer  01 !  02 !  03 !  Total  № Winter  01 !.1  \
Afghanistan (AFG)              13     0     0     2      2         0       0   
Algeria (ALG)                  12     5     2     8     15         3       0   
Argentina (ARG)                23    18    24    28     70        18       0   
Armenia (ARM)                   5     1     2     9     12         6       0   
Australasia (ANZ) [ANZ]         2     3     4     5     12         0       0   

                         02 !.1  03 !.1  Total.1  № Games  01 !.2  02 !.2  \
Afghanistan (AFG)             0       0        0       13       0       0   
Algeria (ALG)                 0       0        0       15       5       2   
Argentina (ARG)               0       0        0       41      18      24   
Armenia (ARM)                 0       0        0       11       1       2   
Australasia (ANZ) [ANZ]       0       0        0        2       3       4   

                         03 !.2  Combined total  
Afghanistan (AFG)             2               2  
Algeria (ALG)                 8              15  
Argentina (ARG)              28              70  
Armenia (ARM)                 9              12  
Australasia (ANZ) [ANZ]       5              12

### Pregunta 1
Cual pais ha ganado mas medallas de oro en los juegos de verano?

In [233]:
df = df.drop('Totals')
df['Total'].idxmax()

'United States\xa0(USA) [P] [Q] [R] [Z]'

### Pregunta 2
Cual pais tiene la diferencia mas grande entre la cantidad de medallas de oro obtenidas en verano e invierno?


In [234]:
df.rename(index={"№ Summer":"spart","01 !":"s01","02 !":"s02","03 !":"s03","Total":"stotal",
               "№ Winter":"wpart","01 !.1":"w01","02 !.1":"w02","03 !.1":"w03","Total.1":"wtotal",
               "№ Games":"totalpart","01 !.2":"total01","02 !.2":"total02","03 !.2":"total03",
               "Combined total":"totalambos"})
df.rename(columns={"№ Summer":"spart","01 !":"s01","02 !":"s02","03 !":"s03","Total":"stotal",
               "№ Winter":"wpart","01 !.1":"w01","02 !.1":"w02","03 !.1":"w03","Total.1":"wtotal",
               "№ Games":"totalpart","01 !.2":"total01","02 !.2":"total02","03 !.2":"total03",
               "Combined total":"totalambos"})

spart  s01  s02  s03  stotal  \
Afghanistan (AFG)                                13    0    0    2       2   
Algeria (ALG)                                    12    5    2    8      15   
Argentina (ARG)                                  23   18   24   28      70   
Armenia (ARM)                                     5    1    2    9      12   
Australasia (ANZ) [ANZ]                           2    3    4    5      12   
...                                             ...  ...  ...  ...     ...   
Yugoslavia (YUG) [YUG]                           16   26   29   28      83   
Independent Olympic Participants (IOP) [IOP]      1    0    1    2       3   
Zambia (ZAM) [ZAM]                               12    0    1    1       2   
Zimbabwe (ZIM) [ZIM]                             12    3    4    1       8   
Mixed team (ZZX) [ZZX]                            3    8    5    4      17   

                                              wpart  w01  w02  w03  wtotal  \
Afghanistan (AFG)                                 0    0    0    0       0   
Algeria (ALG)                                     3    0    0    0       0   
Argentina (ARG)                                  18    0    0    0       0   
Armenia (ARM)                                     6    0    0    0       0   
Australasia (ANZ) [ANZ]                           0    0    0    0       0   
...                                             ...  ...  ...  ...     ...   
Yugoslavia (YUG) [YUG]                           14    0    3    1       4   
Independent Olympic Participants (IOP) [IOP]      0    0    0    0       0   
Zambia (ZAM) [ZAM]                                0    0    0    0       0   
Zimbabwe (ZIM) [ZIM]                              1    0    0    0       0   
Mixed team (ZZX) [ZZX]                            0    0    0    0       0   

                                              totalpart  total01  total02  \
Afghanistan (AFG)                                    13        0        0   
Algeria (ALG)                                        15        5        2   
Argentina (ARG)                                      41       18       24   
Armenia (ARM)                                        11        1        2   
Australasia (ANZ) [ANZ]                               2        3        4   
...                                                 ...      ...      ...   
Yugoslavia (YUG) [YUG]                               30       26       32   
Independent Olympic Participants (IOP) [IOP]          1        0        1   
Zambia (ZAM) [ZAM]                                   12        0        1   
Zimbabwe (ZIM) [ZIM]                                 13        3        4   
Mixed team (ZZX) [ZZX]                                3        8        5   

                                              total03  totalambos  
Afghanistan (AFG)                                   2           2  
Algeria (ALG)                                       8          15  
Argentina (ARG)                                    28          70  
Armenia (ARM)                                       9          12  
Australasia (ANZ) [ANZ]                             5          12  
...                                               ...         ...  
Yugoslavia (YUG) [YUG]                             29          87  
Independent Olympic Participants (IOP) [IOP]        2           3  
Zambia (ZAM) [ZAM]                                  1           2  
Zimbabwe (ZIM) [ZIM]                                1           8  
Mixed team (ZZX) [ZZX]                              4          17  

[146 rows x 15 columns]

In [102]:
(df['stotal']-df['wtotal']).abs().idxmax()


'United States\xa0(USA) [P] [Q] [R] [Z]'

### Pregunta 3

¿Qué país tiene la mayor diferencia entre el recuento de medallas de oro de verano y el recuento de medallas de oro de invierno en relación con el recuento total de medallas de oro?

$$\frac{Summer~Gold - Winter~Gold}{Total~Gold}$$

Solo incluye países que hayan ganado al menos 1 oro tanto en verano como en invierno.

In [154]:
filtro=df[(df.s01>=1) &(df.w01>=1) ] 
ratios = (filtro.s01 - filtro.w01).abs()/filtro.total01
print("El pais con mayor diferencia es: ",ratios.idxmax() + " con un valor de: " +str(ratios.max()))


El pais con mayor diferencia es:  Bulgaria (BUL) [H] con un valor de: 0.9615384615384616


### Pregunta 4

Escribe una función que cree una serie llamada "Puntos", que es un valor ponderado donde cada medalla de oro (`Gold.2`) cuenta 3 puntos, medallas de plata (` Silver.2`) 2 puntos y medallas de bronce (` Bronce.2`) por 1 punto. La función debe devolver solo la columna (un objeto Serie) que creó, con los nombres de los países como índices.

    * Esta función debería devolver una serie denominada `Puntos` de longitud 146 *

In [164]:
import numpy as np
def puntos():
    cant = np.zeros(len(df))
    cant += df.total01 * 3
    cant += df.total02 * 2
    cant += df.total03
    return pd.Series(cant, index=df.index)

In [167]:
cant = puntos()
print(cant)

Afghanistan (AFG)                                 2.0
Algeria (ALG)                                    27.0
Argentina (ARG)                                 130.0
Armenia (ARM)                                    16.0
Australasia (ANZ) [ANZ]                          22.0
                                                ...  
Yugoslavia (YUG) [YUG]                          171.0
Independent Olympic Participants (IOP) [IOP]      4.0
Zambia (ZAM) [ZAM]                                3.0
Zimbabwe (ZIM) [ZIM]                             18.0
Mixed team (ZZX) [ZZX]                           38.0
Name: total01, Length: 146, dtype: float64


## Parte 2

### Pregunta 1

Cargue los datos de energía del archivo Energy Indicators.xls, que es una lista de indicadores de suministro de energía y producción de electricidad renovable de las Naciones Unidas para el año 2013, y debe colocarse en un DataFrame con el nombre de la variable energía.

Tenga en cuenta que este es un archivo de Excel y no un archivo de valores separados por comas. Además, asegúrese de excluir la información del encabezado y el pie de página del archivo de datos. Las dos primeras columnas son innecesarias, por lo que debe deshacerse de ellas y debe cambiar las etiquetas de las columnas para que las columnas sean:

    ['País', 'Suministro de energía', 'Suministro de energía per cápita', '% renovable']

Convierta el suministro de energía en gigajulios (hay 1,000,000 gigajulios en un petajulio). Para todos los países que tienen datos faltantes (por ejemplo, datos con "...") asegúrese de que esto se refleje como valores np.NaN.

Cambie el nombre de la siguiente lista de países (para usar en preguntas posteriores):

    "República de Corea": "Corea del Sur",
    "Estados Unidos de América": "Estados Unidos",
    "Reino Unido de Gran Bretaña e Irlanda del Norte": "Reino Unido",
    "China, Región Administrativa Especial de Hong Kong": "Hong Kong"

También hay varios países con números y / o paréntesis en su nombre. Asegúrese de eliminar estos,

p.ej.

    'Bolivia (Estado Plurinacional de)' debería ser 'Bolivia',

    'Suiza17' debería ser 'Suiza'.



A continuación, cargue los datos del PIB del archivo world_bank.csv, que es un csv que contiene el PIB de los países de 1960 a 2015 del Banco Mundial. Llame a este DataFrame GDP.

Asegúrese de omitir el encabezado y cambie el nombre de la siguiente lista de países:

    "República de Corea": "Corea del Sur",
    "Irán, República Islámica": "Irán",
    "RAE de Hong Kong, China": "Hong Kong"



Finalmente, cargue los datos de Sciamgo Journal y Country Rank para Energy Engineering y Power Technology del archivo scimagojr-3.xlsx, que clasifica los países según sus contribuciones a revistas en el área mencionada anteriormente. Llame a este DataFrame ScimEn.

Únase a los tres conjuntos de datos: PIB, Energía y ScimEn en un nuevo conjunto de datos (utilizando la intersección de los nombres de los países). Utilice solo los datos del PIB de los últimos 10 años (2006-2015) y solo los 15 países principales según el 'Rango' de Scimagojr (Rango 1 a 15).

El índice de este DataFrame debe ser el nombre del país y las columnas deben ser 

['Clasificación', 'Documentos', 'Documentos citables', 'Citas', 'Autocitas', 'Citas por documento', 'H índice ',' Suministro de energía ',' Suministro de energía per cápita ','% Renovable ',' 2006 ',' 2007 ',' 2008 ',' 2009 ',' 2010 ',' 2011 ',' 2012 ',' 2013 ',' 2014 ',' 2015 '].


In [267]:
energia=pd.read_excel('Energy+Indicators.xls')
energia

Unnamed: 0                                         Unnamed: 1  \
0           NaN                                                NaN   
1           NaN                                                NaN   
2           NaN                                                NaN   
3           NaN                                                NaN   
4           NaN                                                NaN   
..          ...                                                ...   
277         NaN                       … denotes no data available.   
278         NaN                                                NaN   
279         NaN                                     Data Quality:    
280         NaN  The data are compiled primarily from the annua...   
281         NaN  For more information, visit http://unstats.un....   

                                            Unnamed: 2 Unnamed: 3 Unnamed: 4  \
0                     Environmental Indicators: Energy        NaN        NaN   
1                                                  NaN        NaN        NaN   
2    Energy Supply and Renewable Electricity Produc...        NaN        NaN   
3                                                  NaN        NaN        NaN   
4                                                  NaN        NaN        NaN   
..                                                 ...        ...        ...   
277                                                NaN        NaN        NaN   
278                                                NaN        NaN        NaN   
279                                                NaN        NaN        NaN   
280                                                NaN        NaN        NaN   
281                                                NaN        NaN        NaN   

                     Unnamed: 5  
0                           NaN  
1                           NaN  
2                           NaN  
3                           NaN  
4    Last update: December 2015  
..                          ...  
277                         NaN  
278                         NaN  
279                         NaN  
280                         NaN  
281                         NaN  

[282 rows x 6 columns]

In [268]:
energia = energia.drop(energia.columns[[0, 1]], axis=1)
energia = energia[17:243]  
energia

Unnamed: 2 Unnamed: 3 Unnamed: 4 Unnamed: 5
17                          Afghanistan        321         10    78.6693
18                              Albania        102         35        100
19                              Algeria       1959         51    0.55101
20                       American Samoa        ...        ...   0.641026
21                              Andorra          9        121    88.6957
..                                  ...        ...        ...        ...
238  Venezuela (Bolivarian Republic of)       2871         95    67.8345
239                            Viet Nam       2554         28    45.3215
240           Wallis and Futuna Islands          0         26          0
241                               Yemen        344         13          0
242                              Zambia        400         26    99.7147

[226 rows x 4 columns]

In [269]:
energia.rename(index={'Unnamed: 2': 'Pais','Unnamed: 3':'Suministro de energía','Unnamed: 4':'Suministro de energía per cápita','Unnamed: 5':'% renovable'},inplace=True)
energia.rename(columns={'Unnamed: 2': 'Pais','Unnamed: 3':'Suministro de energía','Unnamed: 4':'Suministro de energía per cápita','Unnamed: 5':'% renovable'},inplace=True)
energia


Pais Suministro de energía  \
17                          Afghanistan                   321   
18                              Albania                   102   
19                              Algeria                  1959   
20                       American Samoa                   ...   
21                              Andorra                     9   
..                                  ...                   ...   
238  Venezuela (Bolivarian Republic of)                  2871   
239                            Viet Nam                  2554   
240           Wallis and Futuna Islands                     0   
241                               Yemen                   344   
242                              Zambia                   400   

    Suministro de energía per cápita % renovable  
17                                10     78.6693  
18                                35         100  
19                                51     0.55101  
20                               ...    0.641026  
21                               121     88.6957  
..                               ...         ...  
238                               95     67.8345  
239                               28     45.3215  
240                               26           0  
241                               13           0  
242                               26     99.7147  

[226 rows x 4 columns]

In [270]:
energia.replace('...', np.nan, inplace=True)
energia['Suministro de energía'] *= 1000000
energia

Pais  Suministro de energía  \
17                          Afghanistan           3.210000e+08   
18                              Albania           1.020000e+08   
19                              Algeria           1.959000e+09   
20                       American Samoa                    NaN   
21                              Andorra           9.000000e+06   
..                                  ...                    ...   
238  Venezuela (Bolivarian Republic of)           2.871000e+09   
239                            Viet Nam           2.554000e+09   
240           Wallis and Futuna Islands           0.000000e+00   
241                               Yemen           3.440000e+08   
242                              Zambia           4.000000e+08   

     Suministro de energía per cápita  % renovable  
17                               10.0    78.669280  
18                               35.0   100.000000  
19                               51.0     0.551010  
20                                NaN     0.641026  
21                              121.0    88.695650  
..                                ...          ...  
238                              95.0    67.834520  
239                              28.0    45.321520  
240                              26.0     0.000000  
241                              13.0     0.000000  
242                              26.0    99.714670  

[226 rows x 4 columns]

In [271]:
energia.replace({"Pais": {"República de Corea": "Corea del Sur"}},inplace=True)
energia.replace({"Pais": {"Estados Unidos de América": "Estados Unidos"}},inplace=True)
energia.replace({"Pais": {"Reino Unido de Gran Bretaña e Irlanda del Norte": "Reino Unido"}},inplace=True)
energia.replace({"Pais": {"China, Región Administrativa Especial de Hong Kong": "Hong Kong"}},inplace=True)
energia

Pais  Suministro de energía  \
17                          Afghanistan           3.210000e+08   
18                              Albania           1.020000e+08   
19                              Algeria           1.959000e+09   
20                       American Samoa                    NaN   
21                              Andorra           9.000000e+06   
..                                  ...                    ...   
238  Venezuela (Bolivarian Republic of)           2.871000e+09   
239                            Viet Nam           2.554000e+09   
240           Wallis and Futuna Islands           0.000000e+00   
241                               Yemen           3.440000e+08   
242                              Zambia           4.000000e+08   

     Suministro de energía per cápita  % renovable  
17                               10.0    78.669280  
18                               35.0   100.000000  
19                               51.0     0.551010  
20                                NaN     0.641026  
21                              121.0    88.695650  
..                                ...          ...  
238                              95.0    67.834520  
239                              28.0    45.321520  
240                              26.0     0.000000  
241                              13.0     0.000000  
242                              26.0    99.714670  

[226 rows x 4 columns]

In [277]:
energia

Suministro de energía  \
Pais                                                        
Afghanistan                                  3.210000e+08   
Albania                                      1.020000e+08   
Algeria                                      1.959000e+09   
American Samoa                                        NaN   
Andorra                                      9.000000e+06   
...                                                   ...   
Venezuela (Bolivarian Republic of)           2.871000e+09   
Viet Nam                                     2.554000e+09   
Wallis and Futuna Islands                    0.000000e+00   
Yemen                                        3.440000e+08   
Zambia                                       4.000000e+08   

                                    Suministro de energía per cápita  \
Pais                                                                   
Afghanistan                                                     10.0   
Albania                                                         35.0   
Algeria                                                         51.0   
American Samoa                                                   NaN   
Andorra                                                        121.0   
...                                                              ...   
Venezuela (Bolivarian Republic of)                              95.0   
Viet Nam                                                        28.0   
Wallis and Futuna Islands                                       26.0   
Yemen                                                           13.0   
Zambia                                                          26.0   

                                    % renovable  
Pais                                             
Afghanistan                           78.669280  
Albania                              100.000000  
Algeria                                0.551010  
American Samoa                         0.641026  
Andorra                               88.695650  
...                                         ...  
Venezuela (Bolivarian Republic of)    67.834520  
Viet Nam                              45.321520  
Wallis and Futuna Islands              0.000000  
Yemen                                  0.000000  
Zambia                                99.714670  

[226 rows x 3 columns]

In [300]:
GDP=pd.read_csv('world_bank.csv', skiprows=4)
GDP.rename(index={'Country Name': 'Pais'}, inplace=True)
GDP.rename(columns={'Country Name': 'Pais'}, inplace=True)
GDP

Pais Country Code                            Indicator Name  \
0               Aruba          ABW  GDP at market prices (constant 2010 US$)   
1             Andorra          AND  GDP at market prices (constant 2010 US$)   
2         Afghanistan          AFG  GDP at market prices (constant 2010 US$)   
3              Angola          AGO  GDP at market prices (constant 2010 US$)   
4             Albania          ALB  GDP at market prices (constant 2010 US$)   
..                ...          ...                                       ...   
259       Yemen, Rep.          YEM  GDP at market prices (constant 2010 US$)   
260      South Africa          ZAF  GDP at market prices (constant 2010 US$)   
261  Congo, Dem. Rep.          COD  GDP at market prices (constant 2010 US$)   
262            Zambia          ZMB  GDP at market prices (constant 2010 US$)   
263          Zimbabwe          ZWE  GDP at market prices (constant 2010 US$)   

     Indicator Code          1960          1961          1962          1963  \
0    NY.GDP.MKTP.KD           NaN           NaN           NaN           NaN   
1    NY.GDP.MKTP.KD           NaN           NaN           NaN           NaN   
2    NY.GDP.MKTP.KD           NaN           NaN           NaN           NaN   
3    NY.GDP.MKTP.KD           NaN           NaN           NaN           NaN   
4    NY.GDP.MKTP.KD           NaN           NaN           NaN           NaN   
..              ...           ...           ...           ...           ...   
259  NY.GDP.MKTP.KD           NaN           NaN           NaN           NaN   
260  NY.GDP.MKTP.KD  7.707081e+10  8.003449e+10  8.497844e+10  9.124438e+10   
261  NY.GDP.MKTP.KD  1.508024e+10  1.344383e+10  1.629401e+10  1.714358e+10   
262  NY.GDP.MKTP.KD  4.592975e+09  4.655503e+09  4.539542e+09  4.688093e+09   
263  NY.GDP.MKTP.KD  3.338344e+09  3.549199e+09  3.600111e+09  3.824915e+09   

             1964          1965  ...          2006          2007  \
0             NaN           NaN  ...           NaN           NaN   
1             NaN           NaN  ...  4.018196e+09  4.021331e+09   
2             NaN           NaN  ...  1.030523e+10  1.172119e+10   
3             NaN           NaN  ...  5.581103e+10  6.842044e+10   
4             NaN           NaN  ...  9.771760e+09  1.034829e+10   
..            ...           ...  ...           ...           ...   
259           NaN           NaN  ...  2.672565e+10  2.761787e+10   
260  9.848904e+10  1.072507e+11  ...  3.402852e+11  3.585261e+11   
261  1.672524e+10  1.689205e+10  ...  1.650894e+10  1.754232e+10   
262  5.260699e+09  6.136472e+09  ...  1.440569e+10  1.560892e+10   
263  3.782605e+09  3.968352e+09  ...  1.006276e+10  9.695130e+09   

             2008          2009          2010          2011          2012  \
0             NaN           NaN  2.467704e+09           NaN           NaN   
1    3.675728e+09  3.535389e+09  3.346317e+09  3.185605e+09  3.129538e+09   
2    1.214448e+10  1.469733e+10  1.593680e+10  1.691113e+10  1.935220e+10   
3    7.787420e+10  7.975320e+10  8.247091e+10  8.570262e+10  9.012096e+10   
4    1.112752e+10  1.150029e+10  1.192695e+10  1.223109e+10  1.240477e+10   
..            ...           ...           ...           ...           ...   
259  2.872656e+10  2.991436e+10  3.090675e+10  2.624342e+10  2.689160e+10   
260  3.699668e+11  3.642764e+11  3.753494e+11  3.874074e+11  3.960071e+11   
261  1.863448e+10  1.916651e+10  2.052329e+10  2.193213e+10  2.350200e+10   
262  1.682234e+10  1.837342e+10  2.026555e+10  2.140358e+10  2.302438e+10   
263  7.982103e+09  8.459783e+09  9.422161e+09  1.054391e+10  1.165789e+10   

             2013          2014          2015  
0             NaN           NaN           NaN  
1    3.127550e+09           NaN           NaN  
2    1.973134e+10  1.999032e+10  2.029415e+10  
3    9.626143e+10  1.008863e+11  1.039106e+11  
4    1.254247e+10  1.279331e+10  1.312082e+10  
..            ...           ...           ...  
259  2.800914e+10           N

In [301]:
GDP.replace({"Pais": {"República de Corea": "Corea del Sur"}},inplace=True)
GDP.replace({"Pais": {"Irán, República Islámica": "Irán"}},inplace=True)
GDP.replace({"Pais": {"RAE de Hong Kong, China": "Hong Kong"}},inplace=True)
GDP

Pais Country Code                            Indicator Name  \
0               Aruba          ABW  GDP at market prices (constant 2010 US$)   
1             Andorra          AND  GDP at market prices (constant 2010 US$)   
2         Afghanistan          AFG  GDP at market prices (constant 2010 US$)   
3              Angola          AGO  GDP at market prices (constant 2010 US$)   
4             Albania          ALB  GDP at market prices (constant 2010 US$)   
..                ...          ...                                       ...   
259       Yemen, Rep.          YEM  GDP at market prices (constant 2010 US$)   
260      South Africa          ZAF  GDP at market prices (constant 2010 US$)   
261  Congo, Dem. Rep.          COD  GDP at market prices (constant 2010 US$)   
262            Zambia          ZMB  GDP at market prices (constant 2010 US$)   
263          Zimbabwe          ZWE  GDP at market prices (constant 2010 US$)   

     Indicator Code          1960          1961          1962          1963  \
0    NY.GDP.MKTP.KD           NaN           NaN           NaN           NaN   
1    NY.GDP.MKTP.KD           NaN           NaN           NaN           NaN   
2    NY.GDP.MKTP.KD           NaN           NaN           NaN           NaN   
3    NY.GDP.MKTP.KD           NaN           NaN           NaN           NaN   
4    NY.GDP.MKTP.KD           NaN           NaN           NaN           NaN   
..              ...           ...           ...           ...           ...   
259  NY.GDP.MKTP.KD           NaN           NaN           NaN           NaN   
260  NY.GDP.MKTP.KD  7.707081e+10  8.003449e+10  8.497844e+10  9.124438e+10   
261  NY.GDP.MKTP.KD  1.508024e+10  1.344383e+10  1.629401e+10  1.714358e+10   
262  NY.GDP.MKTP.KD  4.592975e+09  4.655503e+09  4.539542e+09  4.688093e+09   
263  NY.GDP.MKTP.KD  3.338344e+09  3.549199e+09  3.600111e+09  3.824915e+09   

             1964          1965  ...          2006          2007  \
0             NaN           NaN  ...           NaN           NaN   
1             NaN           NaN  ...  4.018196e+09  4.021331e+09   
2             NaN           NaN  ...  1.030523e+10  1.172119e+10   
3             NaN           NaN  ...  5.581103e+10  6.842044e+10   
4             NaN           NaN  ...  9.771760e+09  1.034829e+10   
..            ...           ...  ...           ...           ...   
259           NaN           NaN  ...  2.672565e+10  2.761787e+10   
260  9.848904e+10  1.072507e+11  ...  3.402852e+11  3.585261e+11   
261  1.672524e+10  1.689205e+10  ...  1.650894e+10  1.754232e+10   
262  5.260699e+09  6.136472e+09  ...  1.440569e+10  1.560892e+10   
263  3.782605e+09  3.968352e+09  ...  1.006276e+10  9.695130e+09   

             2008          2009          2010          2011          2012  \
0             NaN           NaN  2.467704e+09           NaN           NaN   
1    3.675728e+09  3.535389e+09  3.346317e+09  3.185605e+09  3.129538e+09   
2    1.214448e+10  1.469733e+10  1.593680e+10  1.691113e+10  1.935220e+10   
3    7.787420e+10  7.975320e+10  8.247091e+10  8.570262e+10  9.012096e+10   
4    1.112752e+10  1.150029e+10  1.192695e+10  1.223109e+10  1.240477e+10   
..            ...           ...           ...           ...           ...   
259  2.872656e+10  2.991436e+10  3.090675e+10  2.624342e+10  2.689160e+10   
260  3.699668e+11  3.642764e+11  3.753494e+11  3.874074e+11  3.960071e+11   
261  1.863448e+10  1.916651e+10  2.052329e+10  2.193213e+10  2.350200e+10   
262  1.682234e+10  1.837342e+10  2.026555e+10  2.140358e+10  2.302438e+10   
263  7.982103e+09  8.459783e+09  9.422161e+09  1.054391e+10  1.165789e+10   

             2013          2014          2015  
0             NaN           NaN           NaN  
1    3.127550e+09           NaN           NaN  
2    1.973134e+10  1.999032e+10  2.029415e+10  
3    9.626143e+10  1.008863e+11  1.039106e+11  
4    1.254247e+10  1.279331e+10  1.312082e+10  
..            ...           ...           ...  
259  2.800914e+10           N

In [302]:
GDP.set_index('Pais',inplace=True)

In [303]:
GDP

Country Code                            Indicator Name  \
Pais                                                                      
Aruba                     ABW  GDP at market prices (constant 2010 US$)   
Andorra                   AND  GDP at market prices (constant 2010 US$)   
Afghanistan               AFG  GDP at market prices (constant 2010 US$)   
Angola                    AGO  GDP at market prices (constant 2010 US$)   
Albania                   ALB  GDP at market prices (constant 2010 US$)   
...                       ...                                       ...   
Yemen, Rep.               YEM  GDP at market prices (constant 2010 US$)   
South Africa              ZAF  GDP at market prices (constant 2010 US$)   
Congo, Dem. Rep.          COD  GDP at market prices (constant 2010 US$)   
Zambia                    ZMB  GDP at market prices (constant 2010 US$)   
Zimbabwe                  ZWE  GDP at market prices (constant 2010 US$)   

                  Indicator Code          1960          1961          1962  \
Pais                                                                         
Aruba             NY.GDP.MKTP.KD           NaN           NaN           NaN   
Andorra           NY.GDP.MKTP.KD           NaN           NaN           NaN   
Afghanistan       NY.GDP.MKTP.KD           NaN           NaN           NaN   
Angola            NY.GDP.MKTP.KD           NaN           NaN           NaN   
Albania           NY.GDP.MKTP.KD           NaN           NaN           NaN   
...                          ...           ...           ...           ...   
Yemen, Rep.       NY.GDP.MKTP.KD           NaN           NaN           NaN   
South Africa      NY.GDP.MKTP.KD  7.707081e+10  8.003449e+10  8.497844e+10   
Congo, Dem. Rep.  NY.GDP.MKTP.KD  1.508024e+10  1.344383e+10  1.629401e+10   
Zambia            NY.GDP.MKTP.KD  4.592975e+09  4.655503e+09  4.539542e+09   
Zimbabwe          NY.GDP.MKTP.KD  3.338344e+09  3.549199e+09  3.600111e+09   

                          1963          1964          1965          1966  ...  \
Pais                                                                      ...   
Aruba                      NaN           NaN           NaN           NaN  ...   
Andorra                    NaN           NaN           NaN           NaN  ...   
Afghanistan                NaN           NaN           NaN           NaN  ...   
Angola                     NaN           NaN           NaN           NaN  ...   
Albania                    NaN           NaN           NaN           NaN  ...   
...                        ...           ...           ...           ...  ...   
Yemen, Rep.                NaN           NaN           NaN           NaN  ...   
South Africa      9.124438e+10  9.848904e+10  1.072507e+11  1.120109e+11  ...   
Congo, Dem. Rep.  1.714358e+10  1.672524e+10  1.689205e+10  1.803656e+10  ...   
Zambia            4.688093e+09  5.260699e+09  6.136472e+09  5.794651e+09  ...   
Zimbabwe          3.824915e+09  3.782605e+09  3.968352e+09  4.028795e+09  ...   

                          2006          2007          2008          2009  \
Pais                                                                       
Aruba                      NaN           NaN           NaN           NaN   
Andorra           4.018196e+09  4.021331e+09  3.675728e+09  3.535389e+09   
Afghanistan       1.030523e+10  1.172119e+10  1.214448e+10  1.469733e+10   
Angola            5.581103e+10  6.842044e+10  7.787420e+10  7.975320e+10   
Albania           9.771760e+09  1.034829e+10  1.112752e+10  1.150029e+10   
...                        ...           ...           ...           ...   
Yemen, Rep.       2.672565e+10  2.761787e+10  2.872656e+10  2.991436e+10   
South Africa      3.402852e+11  3.585261e+11  3.699668e+11  3.642764e+11   
Congo, Dem. Rep.  1.650894e+10  1.754232e+10  1.863448e+10  1.916651e+10   
Zambia            1.440569e+10  1.560892e+10  1.682234e+10  1.837342e+10   
Zimbabwe          1.006276e+10  9.695130e+09  7.982103e+09  

In [306]:
ScimEn=pd.read_excel('scimagojr-3.xlsx')
ScimEn

Rank             Country  Documents  Citable documents  Citations  \
0       1               China     127050             126767     597237   
1       2       United States      96661              94747     792274   
2       3               Japan      30504              30287     223024   
3       4      United Kingdom      20944              20357     206091   
4       5  Russian Federation      18534              18301      34266   
..    ...                 ...        ...                ...        ...   
186   187              Guyana          1                  1          0   
187   188    Christmas Island          1                  1          0   
188   189             Reunion          1                  1          2   
189   190         Saint Lucia          1                  1          0   
190   191          Mauritania          1                  1          1   

     Self-citations  Citations per document  H index  
0            411683                    4.70      138  
1            265436                    8.20      230  
2             61554                    7.31      134  
3             37874                    9.84      139  
4             12422                    1.85       57  
..              ...                     ...      ...  
186               0                    0.00        0  
187               0                    0.00        0  
188               1                    2.00        1  
189               0                    0.00        0  
190               0                    1.00        1  

[191 rows x 8 columns]

In [307]:
ScimEn.rename(index={'Country': 'Pais'}, inplace=True)
ScimEn.rename(columns={'Country': 'Pais'}, inplace=True)
ScimEn

Rank                Pais  Documents  Citable documents  Citations  \
0       1               China     127050             126767     597237   
1       2       United States      96661              94747     792274   
2       3               Japan      30504              30287     223024   
3       4      United Kingdom      20944              20357     206091   
4       5  Russian Federation      18534              18301      34266   
..    ...                 ...        ...                ...        ...   
186   187              Guyana          1                  1          0   
187   188    Christmas Island          1                  1          0   
188   189             Reunion          1                  1          2   
189   190         Saint Lucia          1                  1          0   
190   191          Mauritania          1                  1          1   

     Self-citations  Citations per document  H index  
0            411683                    4.70      138  
1            265436                    8.20      230  
2             61554                    7.31      134  
3             37874                    9.84      139  
4             12422                    1.85       57  
..              ...                     ...      ...  
186               0                    0.00        0  
187               0                    0.00        0  
188               1                    2.00        1  
189               0                    0.00        0  
190               0                    1.00        1  

[191 rows x 8 columns]

In [308]:
ScimEn.set_index('Pais', inplace=True)

In [309]:
ScimEn

Rank  Documents  Citable documents  Citations  \
Pais                                                                
China                  1     127050             126767     597237   
United States          2      96661              94747     792274   
Japan                  3      30504              30287     223024   
United Kingdom         4      20944              20357     206091   
Russian Federation     5      18534              18301      34266   
...                  ...        ...                ...        ...   
Guyana               187          1                  1          0   
Christmas Island     188          1                  1          0   
Reunion              189          1                  1          2   
Saint Lucia          190          1                  1          0   
Mauritania           191          1                  1          1   

                    Self-citations  Citations per document  H index  
Pais                                                                 
China                       411683                    4.70      138  
United States               265436                    8.20      230  
Japan                        61554                    7.31      134  
United Kingdom               37874                    9.84      139  
Russian Federation           12422                    1.85       57  
...                            ...                     ...      ...  
Guyana                           0                    0.00        0  
Christmas Island                 0                    0.00        0  
Reunion                          1                    2.00        1  
Saint Lucia                      0                    0.00        0  
Mauritania                       0                    1.00        1  

[191 rows x 7 columns]

In [317]:
df = pd.merge(pd.merge(energia, GDP, on='Pais'), ScimEn, on='Pais')
df

Suministro de energía  Suministro de energía per cápita  \
Pais                                                                            
Afghanistan                    3.210000e+08                              10.0   
Albania                        1.020000e+08                              35.0   
Algeria                        1.959000e+09                              51.0   
American Samoa                          NaN                               NaN   
Andorra                        9.000000e+06                             121.0   
...                                     ...                               ...   
Uganda                         4.520000e+08                              12.0   
United Arab Emirates           2.710000e+09                             300.0   
Uruguay                        1.960000e+08                              58.0   
Uzbekistan                     1.798000e+09                              62.0   
Zambia                         4.000000e+08                              26.0   

                      % renovable Country Code  \
Pais                                             
Afghanistan             78.669280          AFG   
Albania                100.000000          ALB   
Algeria                  0.551010          DZA   
American Samoa           0.641026          ASM   
Andorra                 88.695650          AND   
...                           ...          ...   
Uganda                  68.018180          UGA   
United Arab Emirates     0.000000          ARE   
Uruguay                 71.605040          URY   
Uzbekistan              21.328410          UZB   
Zambia                  99.714670          ZMB   

                                                Indicator Name  \
Pais                                                             
Afghanistan           GDP at market prices (constant 2010 US$)   
Albania               GDP at market prices (constant 2010 US$)   
Algeria               GDP at market prices (constant 2010 US$)   
American Samoa        GDP at market prices (constant 2010 US$)   
Andorra               GDP at market prices (constant 2010 US$)   
...                                                        ...   
Uganda                GDP at market prices (constant 2010 US$)   
United Arab Emirates  GDP at market prices (constant 2010 US$)   
Uruguay               GDP at market prices (constant 2010 US$)   
Uzbekistan            GDP at market prices (constant 2010 US$)   
Zambia                GDP at market prices (constant 2010 US$)   

                      Indicator Code          1960          1961  \
Pais                                                               
Afghanistan           NY.GDP.MKTP.KD           NaN           NaN   
Albania               NY.GDP.MKTP.KD           NaN           NaN   
Algeria               NY.GDP.MKTP.KD  2.744286e+10  2.370914e+10   
American Samoa        NY.GDP.MKTP.KD           NaN           NaN   
Andorra               NY.GDP.MKTP.KD           NaN           NaN   
...                              ...           ...           ...   
Uganda                NY.GDP.MKTP.KD           NaN           NaN   
United Arab Emirates  NY.GDP.MKTP.KD           NaN           NaN   
Uruguay               NY.GDP.MKTP.KD  1.389822e+10  1.424471e+10   
Uzbekistan            NY.GDP.MKTP.KD           NaN           NaN   
Zambia                NY.GDP.MKTP.KD  4.592975e+09  4.655503e+09   

                              1962          1963  ...          2013  \
Pais                                              ...                 
Afghanistan                    NaN           NaN  ...  1.973134e+10   
Albania                        NaN           NaN  ...  1.254247e+10   
Algeria               1.904199e+10  2.557600e+10  ...  1.763249e+11   
American Samoa                 NaN           NaN  ...           NaN   
Andorra                        NaN           NaN  ...  3.127550e+09   
...                            ...           ...  ...           ...   
Ugand

In [314]:
df = pd.merge(pd.merge(energia, GDP, on='Pais'), ScimEn, on='Pais')
df = df[['Rank', 'Documents', 'Citable documents', 'Citations', 'Self-citations', 'Citations per document', 'H index', 'Suministro de energía', 'Suministro de energía per cápita', '% renovable', '2006', '2007', '2008', '2009', '2010', '2011', '2012', '2013', '2014', '2015']]
df = (df.loc[df['Rank'].isin([i for i in range(1, 16)])])
df

Rank  Documents  Citable documents  Citations  \
Pais                                                                
Brazil                15       8668               8596      60702   
Canada                 6      17899              17620     215003   
Germany                7      17027              16831     140566   
India                  8      15005              14841     128763   
Russian Federation     5      18534              18301      34266   

                    Self-citations  Citations per document  H index  \
Pais                                                                  
Brazil                       14396                    7.00       86   
Canada                       40930                   12.01      149   
Germany                      27426                    8.26      126   
India                        37209                    8.58      115   
Russian Federation           12422                    1.85       57   

                    Suministro de energía  Suministro de energía per cápita  \
Pais                                                                          
Brazil                       1.214900e+10                              59.0   
Canada                       1.043100e+10                             296.0   
Germany                      1.326100e+10                             165.0   
India                        3.319500e+10                              26.0   
Russian Federation           3.070900e+10                             214.0   

                    % renovable          2006          2007          2008  \
Pais                                                                        
Brazil                 69.64803  1.845080e+12  1.957118e+12  2.056809e+12   
Canada                 61.94543  1.564469e+12  1.596740e+12  1.612713e+12   
Germany                17.90153  3.332891e+12  3.441561e+12  3.478809e+12   
India                  14.96908  1.265894e+12  1.374865e+12  1.428361e+12   
Russian Federation     17.28868  1.385793e+12  1.504071e+12  1.583004e+12   

                            2009          2010          2011          2012  \
Pais                                                                         
Brazil              2.054215e+12  2.208872e+12  2.295245e+12  2.339209e+12   
Canada              1.565145e+12  1.613406e+12  1.664087e+12  1.693133e+12   
Germany             3.283340e+12  3.417298e+12  3.542371e+12  3.556724e+12   
India               1.549483e+12  1.708459e+12  1.821872e+12  1.924235e+12   
Russian Federation  1.459199e+12  1.524917e+12  1.589943e+12  1.645876e+12   

                            2013          2014          2015  
Pais                                                          
Brazil              2.409740e+12  2.412231e+12  2.319423e+12  
Canada              1.730688e+12  1.773486e+12  1.792609e+12  
Germany             3.567317e+12  3.624386e+12  3.685556e+12  
India               2.051982e+12  2.200617e+12  2.367206e+12  
Russian Federation  1.666934e+12  1.678709e+12  1.616149e+12

In [316]:
df.sort_values("Rank", ascending=True)

Rank  Documents  Citable documents  Citations  \
Pais                                                                
Russian Federation     5      18534              18301      34266   
Canada                 6      17899              17620     215003   
Germany                7      17027              16831     140566   
India                  8      15005              14841     128763   
Brazil                15       8668               8596      60702   

                    Self-citations  Citations per document  H index  \
Pais                                                                  
Russian Federation           12422                    1.85       57   
Canada                       40930                   12.01      149   
Germany                      27426                    8.26      126   
India                        37209                    8.58      115   
Brazil                       14396                    7.00       86   

                    Suministro de energía  Suministro de energía per cápita  \
Pais                                                                          
Russian Federation           3.070900e+10                             214.0   
Canada                       1.043100e+10                             296.0   
Germany                      1.326100e+10                             165.0   
India                        3.319500e+10                              26.0   
Brazil                       1.214900e+10                              59.0   

                    % renovable          2006          2007          2008  \
Pais                                                                        
Russian Federation     17.28868  1.385793e+12  1.504071e+12  1.583004e+12   
Canada                 61.94543  1.564469e+12  1.596740e+12  1.612713e+12   
Germany                17.90153  3.332891e+12  3.441561e+12  3.478809e+12   
India                  14.96908  1.265894e+12  1.374865e+12  1.428361e+12   
Brazil                 69.64803  1.845080e+12  1.957118e+12  2.056809e+12   

                            2009          2010          2011          2012  \
Pais                                                                         
Russian Federation  1.459199e+12  1.524917e+12  1.589943e+12  1.645876e+12   
Canada              1.565145e+12  1.613406e+12  1.664087e+12  1.693133e+12   
Germany             3.283340e+12  3.417298e+12  3.542371e+12  3.556724e+12   
India               1.549483e+12  1.708459e+12  1.821872e+12  1.924235e+12   
Brazil              2.054215e+12  2.208872e+12  2.295245e+12  2.339209e+12   

                            2013          2014          2015  
Pais                                                          
Russian Federation  1.666934e+12  1.678709e+12  1.616149e+12  
Canada              1.730688e+12  1.773486e+12  1.792609e+12  
Germany             3.567317e+12  3.624386e+12  3.685556e+12  
India               2.051982e+12  2.200617e+12  2.367206e+12  
Brazil              2.409740e+12  2.412231e+12  2.319423e+12

### Pregunta 2

La pregunta anterior unió tres conjuntos de datos y luego redujo esto a solo las 15 entradas principales. Cuando se unió a los conjuntos de datos, pero antes de reducir esto a los 15 elementos principales, ¿cuántas entradas perdió?

In [ ]:
#139X69 Paso a 15x21 se perdieron en total 124x48

### Pregunta 3

¿Cuál es el PIB promedio de los últimos 10 años para cada país? (excluya los valores faltantes de este cálculo).

In [329]:
suma=df.loc["Russian Federation":, "2006":"2015"]
suma
avg = suma.sum(axis=1)/10
print (avg)

Pais
Russian Federation      1.565459e+12
Rwanda                  6.000671e+09
Samoa                   6.788420e+08
Senegal                 1.320424e+10
Seychelles              1.044626e+09
Sierra Leone            2.888643e+09
Singapore               2.388876e+11
Slovenia                4.823584e+10
South Africa            3.823935e+11
Sri Lanka               5.997133e+10
Sudan                   6.271866e+10
Suriname                4.432619e+09
Swaziland               3.588678e+09
Sweden                  4.960427e+11
Syrian Arab Republic    0.000000e+00
Tajikistan              5.997592e+09
Thailand                3.454539e+11
Togo                    3.321124e+09
Trinidad and Tobago     2.137886e+10
Tunisia                 4.315751e+10
Turkey                  7.704245e+11
Turkmenistan            2.536781e+10
Uganda                  2.093076e+10
United Arab Emirates    3.102559e+11
Uruguay                 4.056995e+10
Uzbekistan              4.182639e+10
Zambia                  2.057734e